make operator's execution_timeout configurable

sagmansercan opened the pull request:
This PR aims to make the execution_timeout attribute to be configurable globally via airflow.cfg.

* The default value is still `None`. Users are expected to
define an integer value to be passed into timedelta object
to set the timeout in terms of seconds by default, via configuration.
* Added `gettimedelta` method in configuration to be used in abstractoperator
to get timedelta or None type object. The method raises exceptions
for the values that are; 
  * not convertible to integer
  * too large to be converted to C int.
* Sample config cases are added into unit tests.

Closes #18578

sercansagmaninvent committed:
make operator's execution_timeout configurable

By this commit, execution_timeout attribute of the
operators is now configurable globally via airflow.cfg.

* The default value is still `None`. Users are expected to
define a positive integer value to be passed into timedelta object
to set timeout in terms of seconds by default, via configuration.
* If the key is missing or is set to a non-positive value, then it is
considered as `None`.
* Added `gettimedelta` method to be used in abstractoperator
to get timedelta or None type object. The method raises exception
for the values that are not convertible to integer and/or the values
too large to be converted to C int.
* Sample config cases are added into unit tests.

Closes #18578

ashb commented on airflow/configuration.py:
```diff
@@ -608,6 +609,43 @@ def getjson(self, section, key, fallback=_UNSET, **kwargs) -> Union[dict, list,
+
+            # the given value must be positive. Otherwise fallback value is returned.
+            if int_val <= 0:
+                return fallback
```
Why do we need to support negative for none? It feels simpler to have "blank" be None, and negative be an error.

sagmansercan replied:
I thought that it is better not to raise an error for these values and it does not affect the current behavior. However, you're also right. It would be much simpler to not have this logic. I am going to change this, as you've suggested, including zero.

sagmansercan replied:
done ✅

ashb commented on airflow/configuration.py:
```diff
@@ -608,6 +609,43 @@ def getjson(self, section, key, fallback=_UNSET, **kwargs) -> Union[dict, list,
+            except OverflowError:
+                raise AirflowConfigException(
+                    f'Failed to convert value to timedelta in `seconds`. '
+                    f'Please check "{key}" key in "{section}" section. Current value: "{val}".'
```
This error doesn't give the user any clue as to what the error actually is -- as the value it certainly looks like it is a valid number.

sagmansercan replied:
You're right, I'll include the error message.

sagmansercan replied:
The error message is now included in the exception ✅

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -361,6 +361,15 @@
+        The default execution_timeout value for the operators. Expected positive integer value to
+        be passed into timedelta as seconds. The missing and the non-positive values are considered as None,
+        means that the operators are never timed out.
+      version_added: 2.3.1
```
```suggestion
      version_added: 2.3.0
```

We can target 2.3.0 for this one

sagmansercan replied:
done ✅

eladkal commented on airflow/config_templates/config.yml:
```diff
@@ -361,6 +361,15 @@
       type: string
       example: ~
       default: "downstream"
+    - name: default_execution_timeout
```
I think it should be 
```suggestion
    - name: default_task_execution_timeout
```

like: `default_task_retries`, `default_task_weight_rule`, etc...

sagmansercan replied:
Thank you, you're quite right.

sagmansercan replied:
renamed all the configuration parameter and variable names accordingly ✅

sercansagmaninvent committed:
raise error for non-positive execution_timeout

* By this commit, error raises for the values <= 0
instead of using fallback value
* Updated unit tests

sercansagmaninvent committed:
include OverflowError error message in exception

To be more clear to the user, added relevant error message
into to AirflowConfigException.

sercansagmaninvent committed:
rename default_execution_timeout

This parameter specifies the tasks' execution timeout,
so all configuration and variable names are now contains
`task` in it.

sercansagmaninvent committed:
update `version_added` for execution_timeout

SamWheating commented on airflow/configuration.py:
```diff
@@ -608,6 +609,48 @@ def getjson(self, section, key, fallback=_UNSET, **kwargs) -> Union[dict, list,
+
+            # the given value must be positive. Otherwise fallback value is returned.
+            if int_val <= 0:
+                raise AirflowConfigException(
```
Is this error handled anywhere, or will this cause unhandled exceptions? The config reference entry says that:
```
The missing and the non-positive values are considered as None, means that the operators are never timed out.
```

Which suggests that a missing or non-positive value shouldn't raise an error and should instead just return None.

sagmansercan replied:
I've just updated the behavior of this logic based on [this comment](https://github.com/apache/airflow/pull/22389#discussion_r831289387). So, the config reference has become misleading, sorry about that. Let me revisit the explanation again.

sagmansercan replied:
Thanks for the warning. The description is up to date now ✅

sercansagmaninvent committed:
update execution_timeout description

fixed the description of default_task_execution_timeout
based on the recent changes

sercansagmaninvent committed:
update inline comment for non-positive value check

sercansagmaninvent committed:
update `gettimedelta` docstring

pingzh commented on airflow/configuration.py:
```diff
@@ -608,6 +609,48 @@ def getjson(self, section, key, fallback=_UNSET, **kwargs) -> Union[dict, list,
         except JSONDecodeError as e:
             raise AirflowConfigException(f'Unable to parse [{section}] {key!r} as valid json') from e
 
+    def gettimedelta(self, section, key, fallback=None, **kwargs) -> Optional[datetime.timedelta]:
```
nit, wondering if we can call it `get_timedelta` given there are two `t`s.

sagmansercan replied:
hi, this method aims to provide a simple functionality to take a string value from the config file, then try to convert it into an integer, and pass it into timedelta in seconds. So, it is expected to be returned as _a single timedelta object_. It is very similar to the other helper methods like _getint_, _getfloat_, _getboolean_, etc. So I think the method does not support the operation for multiple `t`s.

If I misunderstood your comment, I would like to hear more on this to understand the use cases.

pingzh replied:
hi @sagmansercan i was thinking about the naming of this method, `gettimedelta` vs `get_timedelta`.

`gettimedelta` conforms the current naming convention under `conf`, like `getint`. However, there are two `t`s in the naming, making it harder to read. but it is not a big deal

sagmansercan replied:
Oh, now I see your point :) Yes, it may be a little bit annoying, but as you said, it is compatible with the naming convention in the `configuration.py `. I can apply renaming If everyone is okay with _snake_case_, but for now I am leaving it as it is

uranusjr commented on airflow/configuration.py:
```diff
@@ -608,6 +609,48 @@ def getjson(self, section, key, fallback=_UNSET, **kwargs) -> Union[dict, list,
+                    f'Failed to convert value to timedelta in `seconds`. '
+                    f'Value must be greater than zero. '
+                    f'Please check "{key}" key in "{section}" section. Current value: "{val}".'
+                )
```
Why? timedelta support negative values just fine.

sagmansercan replied:
I thought that in the case of non-positive values, tasks reach the timeout immediately and this did not make sense to me.  So, restricted non-positive values. However, I guess I see your point, users can be held responsible for adjusting this parameter accordingly, then we don't need to handle this in the method, right?

uranusjr replied:
If you want to add a non-negative check, do it outside of `gettimedelta`. I'm fine with having the check, but it's done at the wrong place right now.

sagmansercan replied:
I've removed the integer value check from gettimedelta method. Thank you for your comment @uranusjr, could you please review it again?

sercansagmaninvent committed:
allow non-positive values in gettimedelta

Before this commit, gettimedelta method was preventing
user to provide non-positive values. Now it is totally up to
users to provide a sensible value for this configuration
